### PR TITLE
DPC-1169 implement v2/token

### PR DESCRIPTION
**Pull request by MrMorie**
### Fixes [DPC-1169](https://jira.cms.gov/browse/DPC-1169)

To meet the needs of DPC and follow the Bulk FHIR auth spec, we'll need to combine these capabilities. In particular, we need to provice access tokens to a system after authenticating a request presenting a JWT signed by the public key registered to the system.

To successfully authenticate:

The system ID must be present in the iss and sub claims
The aud claim must match a configurable value (that we expect to set, for instance, to https://sandbox.dpc.cms.gov/api/v2/Token/auth)
The exp and iat claims must be set no more than 5 minutes apart, and presented before expiration
The jti claim must be unique (we suggest a UUID)
The group, system, and public key must not be soft-deleted or expired
The JWT must be signed by the private key corresponding to the registered public key
The JWT must be presented in the body of the request, in the format currently required by DPC and HL7
The Content-Type header must be application/x-www-form-urlencoded
The Accept header must be application/json
To match the DPC system behavior, we need to add "scope": "system/." to the response.

Note: macaroons are not in scope for this ticket, thus we are using the system ID in the iss and sub claims and not macaroons (client tokens) as currently required by DPC.
### Change Details

added v2/token endpoint
added tests
### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [x] security checklist is completed for this change
Security Checklist: https://confluence.cms.gov/pages/viewpage.action?spaceKey=DAPC&title=2021-05-29%3A+SSAS+v2+Security+Checklist

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### CODE REVIEW

**Review by whuang85 — APPROVED** [review]
LGTM

**Review by MrBilnon — DISMISSED** [review]
LATM

File: README.md [hunk 1]
```diff
@@ -100,7 +101,7 @@ Note that to initialize our docker container, we use migrate-and-start, which co
 
 # Test
 
-The SSAS can be tested by running `make test-ssas` or `make unit-test-ssas`. You can also use the repo-wide commands `make test` and `make unit-test`, which will run tests against the entire repo, including the SSAS code.  Some tests are designed to be only run as needed, and are excluded from `make` by a build tag.  To include
+The SSAS can be tested by running `make unit-test`. You can also use the repo-wide commands `make test` and `make unit-test`, which will run tests against the entire repo, including the SSAS code.  Some tests are designed to be only run as needed, and are excluded from `make` by a build tag.  To include
```
Comment (jonfulk) [inline README.md:104]:
```suggestion
The SSAS can be tested by running `make unit-test`. You can also use the repo-wide command `make test`, which will run tests against the entire repo, including the SSAS code.  Some tests are designed to be only run as needed, and are excluded from `make` by a build tag.  To include
```

File: ssas/service/public/tokens.go [hunk 2]
```diff
@@ -3,9 +3,7 @@ package public
 import (
 	"fmt"
 	"time"
-
 	"github.com/dgrijalva/jwt-go"
-
 	"github.com/CMSgov/bcda-ssas-app/ssas"
 	"github.com/CMSgov/bcda-ssas-app/ssas/cfg"
 	"github.com/CMSgov/bcda-ssas-app/ssas/service"
```
Comment (jonfulk) [inline ssas/service/public/tokens.go:9]: did `gofmt` remove these lines? i thought the imports were supposed to be grouped like this.

File: README.md (hunk 1 above)
Comment (MrMorie) [reply README.md:104]: Ty :)

**Review by MrBilnon — DISMISSED** [review]
LGTM

File: ssas/service/server.go [hunk 3]
```diff
@@ -257,6 +256,31 @@ func (s *Server) VerifyToken(tokenString string) (*jwt.Token, error) {
...
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+		claims := token.Claims.(*CommonClaims)
+		if claims.Issuer == ""{
+			return nil, fmt.Errorf("missing issuer (iss) in jwt claims")
+		}
+		system,err := ssas.GetSystemByID(claims.Issuer)
+		if err !=nil {
+			return nil, err
```
Comment (dhgreene) [inline ssas/service/server.go:270]: In the current code `err` is passed to the user unaltered.  Can we make an informative, but generic message either here or in `tokenV2()`?

File: ssas/service/server.go [hunk 4]
```diff
@@ -257,6 +256,31 @@ func (s *Server) VerifyToken(tokenString string) (*jwt.Token, error) {
...
+		}
+		key, err := system.GetEncryptionKey(trackingId)
+		if err !=nil {
+			return nil, fmt.Errorf("key not found for system: %v", claims.Issuer)
+		}
+		pubKey,err := ssas.ReadPublicKey(key.Body)
+		if err !=nil {
+			return nil, err
+		}
+		return pubKey, nil
```
Comment (dhgreene) [inline ssas/service/server.go:280]: Same as above (return a helpful generic message)

File: ssas/service/public/tokens.go (hunk 2 above)
Comment (MrMorie) [reply ssas/service/public/tokens.go:9]: it was go formated

File: ssas/service/public/api.go [hunk 5]
```diff
@@ -541,6 +542,133 @@ func token(w http.ResponseWriter, r *http.Request) {
...
+
+	claims := token.Claims.(*service.CommonClaims)
+	if claims.Subject != claims.Issuer{
+		event.Help = "subject (sub) and issuer (iss) claims do not match"
+		ssas.AuthorizationFailure(event)
+		jsonError(w,"subject (sub) and issuer (iss) claims do not match", "")
+		return
+	}
+
+	if claims.Id ==""{
```
Comment (dhgreene) [inline ssas/service/public/api.go:574]:
```suggestion
	if claims.Id =="" {
```

File: ssas/service/server.go (hunk 4 above)
Comment (MrMorie) [reply ssas/service/server.go:284]: not sure why this does not show up as outdated, but its been updated.

File: ssas/service/public/api_test.go [hunk 6]
```diff
@@ -462,3 +466,444 @@ func (s *APITestSuite) TestJsonError() {
...
+	assert.NotEmpty(s.T(), t)
+	assert.Equal(s.T(), expectedMsg, t.Error)
+}
+
+func mintClientAssertion(issuer string, subject string, aud string, issuedAt int64, expiresAt int64, privateKey *rsa.PrivateKey) (*jwt.Token, string, error) {
+	claims := service.CommonClaims{}
+
+	token := jwt.New(jwt.SigningMethodRS512)
+	tokenID := uuid.NewRandom().String()
+	claims.TokenType = "ClientAssertion"
```
Comment (dhgreene) [inline ssas/service/public/api_test.go:894]: I don't think we're going to have users set the TokenType claim.  Shall we remove this line?

**Review by dhgreene — APPROVED** [review]
💯

**Pull request merged**